packages/vc-verification/src/issuer-verification.ts

```diff
@@ -0,0 +1,146 @@
+    '0x766572694b657900000000000000000000000000000000000000000000000000';
+
+  /**
+   *
+   * @param signer
+   * @param claimManagerAddr
+   */
+  constructor(
+    signer: JsonRpcSigner,
+    roleDefResolverAddr: string,
+    registrySetting: RegistrySettings,
+    ipfsUrl: string
+  ) {
+    this._signer = signer;
+    this._ipfsUrl = ipfsUrl;
+    this._ipfsStore = new DidStore(this._ipfsUrl);
+    this._registrySetting = registrySetting;
+    this._roleDefResolver = RoleDefinitionResolverV2__factory.connect(
+      roleDefResolverAddr,
+      signer
+    );
+    this._didRegistry = EthereumDIDRegistry__factory.connect(
+      registrySetting.address,
+      signer
+    );
+    this._resolver = new Resolver(this._signer.provider, this._registrySetting);
+  }
+
+  /**
+   * Verifies chain of trust for a given holder's credential
+   * @param {string} subjectDID
+   * @param {string} role
+   */
+  async verifyChainOfTrust(subjectDID: string, role: string) {
+    let hasParent = true;
+    while (hasParent) {
+      const subjectDIDDoc = await this._resolver.read(subjectDID);
+      const service = subjectDIDDoc.service;
+      let serviceEndpoint = '';
+      for (const sv of service) {
+        if (utils.namehash(sv.type) === role) {
+          serviceEndpoint = sv.serviceEndpoint;
+          hasParent = false;
+          break;
+        }
+      }
+      const issuerDID = await this.verifyCredential(serviceEndpoint);
+      if (issuerDID) {
+        if (await this.verifyIssuer(role, issuerDID)) {
+          subjectDID = issuerDID;
+        } else {
+          return 'Issuer is not allowed to issue role';
+        }
+      } else {
+        return 'the credential is invalid';
+      }
+    }
+  }
+
+  /**
+   * Verify credential against issuer
+   * @param {string} serviceEndpoint
+   */
+  async verifyCredential(serviceEndpoint: string) {
+    const token = await this._ipfsStore.get(serviceEndpoint);
+    const payload = jwt.decode(token) as { iss: string };
+    const issuerDIDDoc = await this._resolver.read(payload.iss);
+    const verifier = new ProofVerifier(issuerDIDDoc);
+    if (await verifier.verifyAssertionProof(token)) {
+      return payload.iss;
+    } else {
+      return 'Invalid Credential';
+    }
+  }
+
+  /**
+   * Verifies issuer's authority to issue credential
+   * @param {string} role
+   * @param {string} issuerDID
+   * @returns boolean
+   */
+  async verifyIssuer(role: string, issuerDID: string): Promise<boolean> {
+    const { dids, role: string } = await this._roleDefResolver.issuers(
+      utils.namehash(role)
+    );
+    let didMatched = false;
+    if (dids.length > 0) {
+      for (let i = 0; i < dids.length; i++) {
+        if (
+          dids[i] == issuerDID ||
+          (await this._didRegistry.validDelegate(
+            dids[i],
+            this.ASSERTION_DELEGATE_TYPE,
+            issuerDID
+          ))
+        ) {
+          didMatched = true;
+          break;
+        }
+      }
+    }
```

Comment:
I don't think this query of the DID registry should be handled here. It should be handed by the did:ethr resolver (e.g. https://github.com/decentralized-identity/ethr-did-resolver/blob/2a66a2bfc29ebb59a33967f6dec8cdc676c2fee5/src/resolver.ts#L164) and the delegate will just be included in the DID document

Reply:
Ah, I think maybe there is confusion on how delegates for `did:ethr`. I could be wrong, but as I see it delegates in `did:ethr` are addresses that are allowed to sign on behalf of the DID, they are not addresses that are equivalent identifiers to the DID.

In other words, if the issuer of a credential is a delegate of a valid issuer according to the role definition, this does not make the issuer of the credential valid.

For example, if the role definition specifies that `did:123` the valid issuer, then this is a valid credential:
```js
{
    issuer: "did:123",
    credentialSubject: { ... },
    proof: {
      verificationMethod: "did:123#delegate-did:456"
      ....
   }
```
whereas this is not:
```js
{
    issuer: "did:456",
    credentialSubject: { ... },
    proof: {
      verificationMethod: "did:456#controller"
      ....
   }
```

In other words, if did:456 wants to issue a credential on behalf of did:123, we should expect that did:456 should still put did:123 as the `issuer` in the credentials that they are issuing.
This is more correct in my view and has the benefit of allowing the code to remain generic to the general DID spec and not coupled to did:ethr in anyway.